ROX-13904: Excent plop lifecycle with closed status [Do not merge, used for Pipeline testing]

erthalion opened the pull request:
## Description

Derive "closed" status for ProcessListeningOnPort and modify the lifecycle accordingly. Closed PLOP objects are going to be ignored from the API.

## Checklist
- [ ] Investigated and inspected CI test results
- [x] Unit test and regression tests added

## Testing Performed

Local unit tests run.

erthalion committed:
Add namespace to the storage api

erthalion committed:
Pruning for PLOP

Take into account ProcessListeningOnPort when pruning ProcessIndicators.
They have an FK corresponding to a relation between those two entities,
not enforced by the db though. To prevent orphaned PLOP objects, clean
them up together.

JoukoVirtanen committed:
Public API implementation for PLOP

Implement service for public v1 API corresponding to
ProcessListeningOnPort.

erthalion committed:
Extend data store

Add group by implementation for select by namespace, introduce necessary
metrics and SAC.

Co-authored-by: Jouko Virtanen <jvirtane@redhat.com>

erthalion committed:
Fix process pruning

Always provide real store when needed, and generate OptionsMap required
for DeleteByQuery.

erthalion committed:
Turn processindicator fk into UUID

It was already transformed in the main table, the fk has to be using
UUID as well.

JoukoVirtanen committed:
Update e2e tests for PLOP

Update & fix for e2e ProcessesListeningOnPortsTest

erthalion committed:
Disable generation checks

Disable generation for ProcessListeningOnPorts and ProcessIndicator to
keep static checks green. This change is temporary until store generator
will support those changes.

erthalion committed:
Add forgotten mocks

Generated mocks are missing in git tree.

JoukoVirtanen committed:
Guards for PLOP to run only on PG

ProcessListeningOnPort is a PostgreSQL feature-only. Add corresponding
checks and tests.

Co-authored-by: Dmitrii Dolgov <ddolgov@redhat.com>

erthalion committed:
Add PLOP gen file with schema-only option

erthalion committed:
Store vs DataStore naming

Be more careful with difference in naming and comments

erthalion committed:
Verify presense of the process information

It's not clear whether such cases are possible, but the protective style
is always a good idea.

erthalion committed:
Type in the name

Fix type in the singleton.go file name

erthalion committed:
Initalize results with proper size value

Result variable is carrying transformed version of PLOP list, so it
makes sense to make the slice of the corresponding size.

erthalion committed:
Temporary downgrade generated files to normal

Clarify that PLOP store and ProcessIndicator schema are not generated
for time being. It comes with some formatting for the former file.

JoukoVirtanen committed:
E2E tests for PLOP

Base implementation for integration tests.

erthalion committed:
Use DeploymentExtension SAC

ProcessListeningOnPort SAC resource isn't needed and in the absense of
hierarchy of dependencies between resources pick up DeploymentExtension
as other similar objects do.

JoukoVirtanen committed:
X-Smart-Squash: Squashed 9 commits:

f71b90802f Added cluster_id to protobuf
29a89c7a44 Moved location of cluster_id in proto definitions
1e7d175a95 Added debugging
b53dbc6578 More debugging
8dac5323ec More debugging
e5b54d5da3 Getting cluster_id from query
5f0780a480 Fixed style errors
da89b5d58a Fixed some errors
4aeb238a27 Updated e2e tests to check for cluster_id

JoukoVirtanen committed:
X-Smart-Squash: Squashed 8 commits:

c54b167c37 Working on changing API
037d07b502 Trying to run make go-generated-srcs
42e9fdfa21 Updated pkg/metrics/operations.go
c21b8f80a6 Fixed query
7e52d00fac Test passes
39c7762ad1 Cleaned up tests
dd1f320a2d Added asserts for second deployment
5bf244fc5e Fixed minor style error

erthalion committed:
Add PLOP insert tests

Simple tests to verify that insertion happened without errors, with a
couple of corner cases as well.

JoukoVirtanen committed:
X-Smart-Squash: Squashed 23 commits:

0a2069ce32 Working on adding ProcessSignal to process listening on port
2c73957bd1 Able to build image
1c3c8cc965 Fixed typo
ca07ce2112 Api works but many fields are blank
74bd3844cf Trying to get pid
df50c26c7e Trying to use serialized from process indicator
107e1c77b9 Removed pid
8b1c8494bf Trying to get all fields from proc serialized
46f157b297 Restored to when pid was removed
f2376b38b1 Trying to add pid
f65cfd079a Trying to add signal id
feecb85d4d Added uid
bd37b55729 Added gid
056504b857 Added scraped
cac533fa45 Added lineage
62b12a20ae Added lineage info
9c684ba352 Cleaned up
6c3a67e706 More cleaning
b69faf2764 Tests pass but don't check all fields yet
b8b20087ad Fixed style errors
31195f0b4c make go-generated-srcs
8ce2c095e8 Fixed style check
c377545fc1 Added asserts for new attributes to e2e tests

JoukoVirtanen committed:
Fixed style error

erthalion committed:
Add more verbose dev docs.

JoukoVirtanen committed:
X-Smart-Branch-Parent: ddolgov/feature/process-listening-on-port

JoukoVirtanen committed:
Moved port and protocol to endpoint struct

JoukoVirtanen committed:
X-Smart-Branch-Parent: jv-rox-13774-add-cluster_id-to-processes-listening-on-ports-api

JoukoVirtanen committed:
X-Smart-Branch-Parent: jv-rox-13823-process-listening-on-ports-api-should-take-only-the-deployment-as-a-parameter

JoukoVirtanen committed:
X-Smart-Branch-Parent: jv-rox-13826-processes-listening-on-ports-api-should-return-process-signal

erthalion committed:
Adjust unit tests

Fix unit tests for PLOP data storage to reflect changes made in the
previous commits about separating a public API PLOP object from an
internal API version.

JoukoVirtanen committed:
X-Smart-Squash: Squashed 2 commits:

03b9338d83 Added missing fields
b5fe5f1c4f Removed commented out code

erthalion committed:
Add Closed column

Add new boolean column to PLOP to identify if the listening was finished
or not. In this way we introduce only one small size column, instead of
basing this on full ClosedTimestamp.

erthalion committed:
Update generated PLOP store

Run code generation again to incorporate GetByQuery needed for the
following commits to implement closing lifecycle for PLOP.

erthalion committed:
Implement closing lifecycle for PLOP

PLOP objects should be distinguished to "active" and "closed" based on
the ClosedTimestamp value. If it's set to not null value, it means the
event we receive notifies about the endpoint beint closed. In that case
instead of creating a new record, update the existing one to be closed.

Such closed events are going to be ingored from the API.

erthalion committed:
Add tests for closing lifecycle

Add more unit test to verify different cases of closing PLOP object,
i.e. happy path (open/close), the case when there was no active PLOP
reported. Verify that it's getting excluded from the API.

NOTE: This implementation doesn't not handle cases when open/close PLOP
are in the same batch.

JoukoVirtanen commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -105,6 +107,40 @@ func (ds *datastoreImpl) AddProcessListeningOnPort(
+		// XXX: This has to be done in a single join query fetching both
+		// ProcessIndicator and needed bits from PLOP.
+		existingPLOP, err := ds.storage.GetByQuery(ctx, search.NewQueryBuilder().
+			AddStrings(search.ProcessID, indicatorIds...).
```
Above we are using DisjunctionQuery. Here we are using an array of indicatorIds. We could use DisjuctionQuery here or we could use an array of indicatorIds above were we use DisjunctionQuery. What are the advantages and disadvantages of the methods? Why use these two methods? I don't know that one is better than the other.

JoukoVirtanen commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -127,15 +163,40 @@ func (ds *datastoreImpl) AddProcessListeningOnPort(
+			}
+		} else {
+			plopObjects[i] = newPLOP
+		}
```
This is a very long function. Maybe break it into smaller pieces.

JoukoVirtanen commented on proto/storage/process_listening_on_port.proto:
```diff
@@ -50,4 +50,7 @@ message ProcessListeningOnPortStorage {
         [(gogoproto.moretags) = 'search:"Process ID,store" sql:"fk(ProcessIndicator:id),no-fk-constraint,index=btree,type(uuid)"'];
+
+    // XXX: Make it a partial index on only active, not closed, PLOP
+    bool closed = 6 [(gogoproto.moretags) = 'search:"Closed,store"  sql:"index=btree"'];
```
I don't think you need this additional field. The close_timestamp already keeps track of if the port is open or closed. If close_timestamp is nil the port is open. If close_timestamp is not nil the port is closed.

erthalion replied:
`close_timestamp` is part of serialized field and is not extracted into a separate column to operate via sql. I was thinking about this option, but decided to use a boolean value instead because we have to create an index on it, and in this way it would be smaller than an index over a timestamp field (especially when partial index support will be introduced into the generator).

JoukoVirtanen commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -127,15 +163,40 @@ func (ds *datastoreImpl) AddProcessListeningOnPort(
+		}
 	}
 
 	// Now save actual PLOP objects
```
I think this is overly complicated. I don't think you needed to make so many changes to the existing code that you had. I don't think this Closed field is needed. I commented on this below in the protobuf definition file, but for convenience I will include my comment there, here. 

"I don't think you need this additional field [closed]. The close_timestamp already keeps track of if the port is open or closed. If close_timestamp is nil the port is open. If close_timestamp is not nil the port is closed."

It seems like all of this code is here just to set the Closed field and that field is not needed.I think probably all that needs to be done is to modify the query in store.go so that it only returns entries where close_timestamp is nil. We are using upsert so we should be able to insert new plops without worrying if they already exist.

erthalion replied:
> It seems like all of this code is here just to set the Closed field and that field is not needed.I think probably all that needs to be done is to modify the query in store.go so that it only returns entries where close_timestamp is nil. We are using upsert so we should be able to insert new plops without worrying if they already exist.

Not quite, most of the code is busy updating closed_timestamp instead of creating another duplicating record of the PLOP object with different closed_timestamp. To achieve that there should be extra lookup (or ideally an extension of process indicator with the join, but it's more cumbersome with the current store generator) to fetch already existing potentially fitting PLOP objects (as neither Collector nor Sensor know about PLOP id) -- this is what constitutes most of the code changes.

JoukoVirtanen replied:
Hmm. I might not be clear on how upsert works in this case. If we do an upsert with different close_timestamps, why do we get duplicates, if we don't get duplicates when upserting different closed values? I might need to play around with this. I have read about upsert in postgres, but I don't know exactly what the UpsertMany function does.

erthalion replied:
> If we do an upsert with different close_timestamps, why do we get duplicates

The `UPSERT` in Postgres relies on a unique constraint to verify, if the constraint is failing -- the update will be performed, otherwise the insert. The storage generator uses an id unique constraint for this purpose, which means:

* If we call `upsert` on an object with `id` field set and there is already existing record with this value -- an actual upsert is happening
* If we call `upsert` on an object without `id` set, there will be no clash possible and it will be always inserted.

The data Central received from Sensor cannot contain a PLOP `id`, because only Central knows it. Thus  there should be fetching logic. Does it make things clearer to you?

JoukoVirtanen replied:
Yes. I was thinking that the key used for the upsert would be process_indicator_id, but I see that it is not the primary key and id is the primary key. Could we use process_indicator_id as the primary key?

erthalion replied:
> Could we use process_indicator_id as the primary key?

Not really, this will break cases when one process listens to multiple ports.

JoukoVirtanen replied:
You can also specify a column to be unique. E.g. in storage/cluster.proto there is the following 

message Cluster {
    string                    id                               = 1 [(gogoproto.moretags) = 'search:"Cluster ID,hidden,store" sql:"pk,type(uuid)"'];
    string                    name                             = 2 [(gogoproto.moretags) = 'search:"Cluster,store" sql:"unique"'];

Perhaps process_indicator_id could be set to be unique. Would that simplify things?

erthalion replied:
> Perhaps process_indicator_id could be set to be unique. Would that simplify things?

As I've already said, no, this will break cases when one process listens to multiple ports. You have to keep in mind that the relation here is "one to many", i.e. one process to many ports.

dashrews78 replied:
I see both sides of this conversation.  This is a lot of logic to insert a record at the same time, based on how we set up the table it is necessary.  I do wonder, though, would it be feasible to change to a composite key and take advantage of that?  Say we eliminate the plop ID itself and got with a composite key on indicator_id, port, and protocol.  Something like:

`message ProcessListeningOnPortStorage {
    // Ideally it has to be GENERATED ALWAYS AS IDENTITY, which will make it a
    // bigint with a sequence. Unfortunately at the moment some bits of store
    // generator assume an id has to be a string.
    string id = 1 [(gogoproto.moretags) = 'sql:"pk,type(uuid)"'];

    uint32 port = 2  [(gogoproto.moretags) = 'sql:"pk"`;
    L4Protocol protocol = 3 [(gogoproto.moretags) = 'sql:"pk"`;
    google.protobuf.Timestamp close_timestamp = 4;

    string process_indicator_id = 5
        [(gogoproto.moretags) = 'search:"Process ID,store" sql:"pk,fk(ProcessIndicator:id),no-fk-constraint,index=btree,type(uuid)"'];

    // XXX: Make it a partial index on only active, not closed, PLOP
    bool closed = 6 [(gogoproto.moretags) = 'search:"Closed,store"  sql:"index=btree"'];
}`

I think the generators would support that and it may help simplify some of this logic.  That said, my background on PLOP is high level so there may be reasons that what I suggest is not feasible.

erthalion replied:
It's an interesting idea, I haven't though about this one -- in this way it doesn't make sense to keep anything serialized anymore so we can unwrap it. There are of course limitations as well:

* In this case we loose possibility to store those PLOP objects what we were not able to correlate with any process information. There could be cases like this, so the idea was to still keep them to not loose the data and for introspection about why did it happen like this.
* The primary index is going to be rather large with certain amount of duplication as endpoint information would be often repeated. The proposal is essentially to trade one read query for an index over three columns. It would be interesting to see if PG14 posting lists could optimize this case.

The amount of changes required to implement this is comparable to the original approach, pruning and custom store parts have to be changed.

erthalion replied:
> This is a lot of logic to insert a record

Btw, for both to keep in mind: the whole acrobatics around existing PLOP map is needed only because it's hard to fetch the indicator + plop info in one join. I've started doing this, but already tired of fighting with it's limitations and the custom version would be annoying as well because of batching. But as soon as joins will be normally supported by query builder this whole patch will become:

* fetch indicator and correlated plop if exists
* based on the presence of the latter fill up the structure to upsert with either existing id or with the new one.

erthalion replied:
> all these options have drawbacks. IMO if the generators and query builder are holding you back, it might be time to tackle those updates that you need even though I know you would rather not do that as part of MVP.

It would be nice, but we've already spent much more than expected time on this, so not sure.

 > There could be cases like this, so the idea was to still keep them to not loose the data and for introspection about why did it happen like this.

That reminds me, I think we've forgotten to actually keep needed process information in this case in the serialized column :( I guess there should be a process info field, that will contain the signal details we've got from Sensor which were not found in the db, and null otherwise.

JoukoVirtanen commented on central/processlisteningonport/datastore/datastore_impl_test.go:
```diff
@@ -254,11 +467,24 @@ func (suite *PLOPDataStoreTestSuite) TestPLOPAddMultipleIndicators() {
+			Args:         "test_arguments1",
+			ExecFilePath: "test_path1",
+		},
+	})
```
Unit tests look good. Could you try writing datastore_impl.go without the closed field and without getting existing plops and see if the unit tests pass?

JoukoVirtanen commented on central/processlisteningonport/store/postgres/store.go:
```diff
@@ -57,8 +57,7 @@ const (
 		"ON plop.processindicatorid = proc.id " +
-		"WHERE proc.deploymentid = $1"
-
+		"WHERE proc.deploymentid = $1 AND plop.closed = false"
```
Try checking if close_timestamp is null instead.

dashrews78 commented on central/processindicator/datastore/datastore_impl.go:
```diff
@@ -137,12 +142,28 @@ func (ds *datastoreImpl) removeMatchingIndicators(ctx context.Context, results [
 		return err
 	}
+
+	if env.PostgresDatastoreEnabled.BooleanSetting() {
```
Keep an eye on this in scale and performance tests.  I tried some similar things with process indicators and network flows but in those scale environments I would get statement timeouts on these big deletes.  That is why DeleteMany batches.  It may be OK here, just something to keep in mind when testing.

erthalion replied:
Good point, thanks. Generally speaking I believe it would be much better to rework pruning to do the cleanup by dropping old partitions, but it's quite a lot of work.

I've planned one round of perf evaluation for sql queries, so will check this out.

erthalion committed:
Implement pruning for PLOP

There are chances to have PLOP objects stored in the db without properly
setup reference to a Process Indicator. They're not getting ignored to
not lose the data and for introspection purposes, but constantly
aggregating them is not possible. Thus extend pruning to clean up such
objects following the regular pruning rules (most importantly
orphanWindow). The implementation follow the established suboptimal
approach and has to be reworked in the future.

erthalion committed:
Extend readme

Add lifecycle info and metrics.

erthalion committed:
Add ProcessIndicatorUniqueKey

Store ProcessIndicator information in the serialized column in case if
we were not able to find matching ProcessIndicator in the database. In
this way we do not lose information if something goes wrong, but do not
do duplication in case if the reference PLOP to ProcessIndicator was
established correctly.

erthalion committed:
Move port and protocol fields out of serialized

There are two cases when we would like to be able to operate with port
and protocol fields on the database side:

* Doing queries like "which processes listen on port X"

* Adding where clauses into the existing queries to minimize amount of
data passed between the db and the application

To enable this, move out those two fields out of opaque serialized
column into corresponding separated columns. This would allow us to be
more flexible, without needing to perform any kind of schema and data
migrations in the future to support the cases mentioned before.

erthalion committed:
Add batch normalization

The PLOP events batch can contain multiple events per port, protocol and
process. Those events have to be treated as a single event, because it's
consistent with how such situations handled elsewhere in the system (on
the Collector and DB side).

erthalion committed:
Allow PLOP reopening

Multiple event pairs open/close should be treated as a single event,
updating CloseTimestamp when needed. Eventually PLOP record in the
database should represent only the current state of things, not the
historical records.

erthalion committed:
Refactor AddProcessListeningOnPort

erthalion committed:
Update protolock file

erthalion committed:
Update PLOP datastore mocks

erthalion commented:
/retest-required

JoukoVirtanen committed:
Fixed style error in ProcessesListeningOnPortsTest.groovy by removing bad import

JoukoVirtanen committed:
Removed another import

JoukoVirtanen committed:
Fixed style-check errors

erthalion committed:
Prune PLOPs only for Postgres

JoukoVirtanen committed:
Add tests for batch processing

Concentrace on corner cases: events completed inside the batch, ourder
of events, etc.

erthalion committed:
Batch normalization improvements

Fix various corner cases for batch normalization:

* Avoid situations when adding nil plop object (do not allocate upfront,
append instead)
* Separate normalized events into normal and those that are completed
inside the batch
* Sort event inside the batch by CloseTimestamp for out-of-order events

erthalion committed:
Process Listening On Ports

Implement data store needed to persist and query PLOP objects, and
corresponding internal API pipeline to actually process it.
